website/src/app/blog/sans-io/readme.mdx

```diff
@@ -0,0 +1,558 @@
+As a result, it is very easy to compose these state machines together: Want to query 5 STUN servers for your public IP? No problem.
+Just make 5 `StunBinding`s and call them in order[^3].
+
+In the case of firezone, you can see this in the example of [`snownet`](https://github.com/firezone/firezone/tree/main/rust/connlib/snownet), a library that combines ICE & wireguard and thereby exposes "magic" IP tunnels that work in any network setup to the rest of the application.
+
+`snownet` builds on top of `str0m`, a sans-IO WebRTC library and `boringtun`, an (almost[^2]) sans-IO wireguard implementation.
+We don’t need the majority of the WebRTC stack though.
+The only thing we are interested in is the `IceAgent` which implements [RFC 8845](https://github.com/cloudflare/boringtun/issues/391).
+ICE uses a clever algorithm that ensures two agents, deployed into arbitrary network environments find the most optimal communication path to each other.
+The result of ICE is a pair of socket addresses that we then use to perform a wireguard handshake.
+Because `str0m` is built in a sans-IO fashion, only using the `IceAgent` part of it is shockingly trivial:
+you simply only import that part of the library and compose its state machine into your existing code.
+In `snownet`, a [connection](https://github.com/firezone/firezone/blob/a5b7507932e9d27e3fc9ed5be7428b9937f2f828/rust/connlib/snownet/src/node.rs#L1289-L1306) simply houses an `IceAgent` and a wireguard tunnel, dispatching incoming messages to either one or the other.
+
+### Flexible APIs
+
+sans-IO code needs to be "driven" by an eventloop of some sorts because it "just" expresses the state of the system but doesn’t cause any side-effects itself.
+The eventloop is responsible for "querying" the state (like `poll_transmit`), executing it and also passing new input to the state machine (`handle_timeout` and `handle_input`).
+To some people, this may appear as unnecessary boilerplate but it comes with a great benefit: flexibility.
+
+* Want to make use of `sendmmsg` to reduce the number of syscalls when sending packets? No problem.
+* Want to multiplex multiple protocols over a single socket? No problem.
+
+Writing the eventloop yourself is an opportunity to be able to tune our code to exactly what we want it to do.
+This also makes maintenance easier for library authors: They can focus on correctly implementing protocol functionality instead of having debates around async runtimes or exposing APIs to set socket options.
+
+A good example here is `str0m`’s stance on the gathering of candidates:
+This is an IO concern and up to the application on how to achieve it.
+`str0m` only provides an API to add such an ICE candidate to the current state.
+As a result, we are able to easily implementation optimisations such as gathering TURN candidates prior to any connection being made, thus reducing firezone's connection-setup latency.
+
+### Testing at the speed of light
```

Comment:
Would it be worth expanding on this section a bit more about the issues we face testing Firezone in CI? Or save that for another article?

We could do a part 2 or something where we detail our CI/CD architecture and how we are chipping away at testing closer and closer to the damn socket from either side of the IO boundary.

Reply:
> Would it be worth expanding on this section a bit more about the issues we face testing Firezone in CI? Or save that for another article?
> 
> We could do a part 2 or something where we detail our CI/CD architecture and how we are chipping away at testing closer and closer to the damn socket from either side of the IO boundary.

I think it is worth doing an entire article on "How we test Firezone". I am only touching on this topic here as a benefit of sans-IO. Talking more about testing seems off-topic.